### PR TITLE
[DF] Add GetDataSourceInfo function

### LINKED ISSUE #7828: [DF] Introduce GetDataSourceInfo and add info to Describe
The idea is to have an identifier that describes the used datasource (or TTree or simple DF), make this info accessible as public API and also add it to `Describe`. See the discussion here:

https://github.com/root-project/root/issues/7561#issuecomment-817629082

https://github.com/root-project/root/issues/7561#issuecomment-817717882

Remember to make the identifier returned by this API subject to change because it's for interactive use and not to be parsed automatically.

### PR DESCRIPTION
**Pull request by stwunsch**
Fixes #7828 partially, missing integration into Describe

### CODE REVIEW

**Comment by vepadulano** [comment]
I think it's the right way to go, maybe the ttree datasource could be further specialized to distinguish between TTree and TChain.
But, a more notable absence in this logic is horizontal extensions to the dataset, e.g. friends in the case of TTree.

So if we imagine this function to be just "giving an hint about the type of container for the dataset of this RDataFrame", then it's already good as it is imho.

If we want something more detailed, like "this dataframe will run its computations on a dataset named XXX, stored in a container of type YYY placed at ZZZ", then we're looking for more. For a ttree without friends, this would translate to
* XXX = name of the tree
* YYY = ROOT file / TFile
* ZZZ = path to the ROOT file
And for a csv table it would be
* XXX = `None`
* YYY = CSV file
* ZZZ = path to the csv file

If the dataset is horizontally extended through e.g. friends, then the information above is not sufficient to identify the full dataset RDataFrame will run its computations on. For a ttree, we would also need to print out XXX,YYY,ZZZ for any friend of the tree. So what I'm saying is that we either give *very high level* info about data (which I would prefer), or we have to go really deep down to be consistent with any datasource type

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 1]
```diff
@@ -2106,6 +2106,28 @@ public:
...
+   ///
+   /// Example usage:
+   /// ~~~{.cpp}
+   /// ROOT::RDataFrame df("tree", "file.root");
+   /// std::cout << df.GetDataSourceInfo() << std::endl;
+   /// // prints "TTree datasource"
+   /// ~~~
+   std::string GetDataSourceInfo() const {
+      if (fLoopManager->GetTree()) {
+         return "TTree datasource";
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2123]: could be a TChain

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 2]
```diff
@@ -2106,6 +2106,28 @@ public:
...
+      } else if (fDataSource) {
+         return fDataSource->GetLabel();
+      } else {
+         return "trivial datasource";
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2127]: "No datasource" would be clearer, or even better "No datasource, X events"

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 3]
```diff
@@ -2106,6 +2106,28 @@ public:
    /// ~~~
    unsigned int GetNRuns() const { return fLoopManager->GetNRuns(); }
 
+   /// \brief Gets descriptive information about the datasource
+   /// \return Info describing the datasource
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2110]:
```suggestion
   /// \return A string describing the kind of datasource in use.
```

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 4]
```diff
@@ -2106,6 +2106,28 @@ public:
...
+   /// The information returned by this convenience function is meant for interactive
+   /// use. The exact string format should not be parsed automatically and can be subject to change.
+   ///
+   /// Example usage:
+   /// ~~~{.cpp}
+   /// ROOT::RDataFrame df("tree", "file.root");
+   /// std::cout << df.GetDataSourceInfo() << std::endl;
+   /// // prints "TTree datasource"
+   /// ~~~
+   std::string GetDataSourceInfo() const {
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2121]: I'm now thinking that "Info" is too generic, what do you think of `GetDataSourceKind`?

**Review by eguiraud — COMMENTED** [review]
Thanks a lot @stwunsch

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx (hunk 4 above)
Comment (vepadulano) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2121]: `DataSourceKind` is good, but I think `GetDataFormatName` better reflects the return value of this function. I see it as being parallel to the recent change in the rdf docs https://root.cern/doc/master/classROOT_1_1RDataFrame.html#other-file-formats

Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2121]: Eh, names are hard... TTrees and TChains are the same data format. "No datasource" is not really a data format.

Comment (vepadulano) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2121]: True, in fact I'd return "TTree" data format for both TTree and TChain. For the empty RDF with entries the semantic it's different, I agree.

Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2121]: I think this method would be more useful if it distinguished between TTree and TChain ([as you also mentioned](https://github.com/root-project/root/issues/7828#issuecomment-817820448))

**Comment by stwunsch** [comment]
We agreed that we want this alltogether in #7833. Closing here for now!

**Pull request closed without merging**